File: infrastructures/rest-ddd/typescript/src/domain/models/BaseModel.ts
/**
 * Base Model (BMO) - Business Model Object
 * Pure business logic with NO database attributes.
 * All domain models MUST extend this base class.
 */

export abstract class BaseModel {
  /**
   * Abstract base class for all Business Model Objects (BMOs).
   * 
   * Rules:
   * - NO database/persistence attributes
   * - Contains business logic and invariants
   * - Validates business rules
   * - Immutable by default (use readonly properties)
   */

  /**
   * Validate business invariants.
   * Throws exception if invalid.
   */
  abstract validate(): void;

  /**
   * Convert model to plain object representation.
   * Used by mappers for transformation.
   */
  abstract toObject(): Record<string, any>;
}

File: infrastructures/rest-ddd/python/src/domain/models/base_model.py
"""
Base Model (BMO) - Business Model Object
Pure business logic with NO database attributes.
All domain models MUST inherit from this base class.
"""
from abc import ABC, abstractmethod
from typing import Any, Dict


class BaseModel(ABC):
    """
    Abstract base class for all Business Model Objects (BMOs).
    
    Rules:
    - NO database/persistence attributes
    - Contains business logic and invariants
    - Validates business rules
    - Immutable by default (use frozen dataclasses or properties)
    """
    
    @abstractmethod
    def validate(self) -> None:
        """
        Validate business invariants.
        Raises exception if invalid.
        """
        pass
    
    @abstractmethod
    def to_dict(self) -> Dict[str, Any]:
        """
        Convert model to dictionary representation.
        Used by mappers for transformation.
        """
        pass
